src/sentry/services/hybrid_cloud/__init__.py

```diff
@@ -226,6 +229,47 @@ def __getattr__(self, item: str) -> Any:
         raise KeyError(f"No implementation found for {cur_mode}.")
 
 
+class DelegatedByOpenTransaction(Generic[ServiceInterface]):
+    """
+    It is possible to run monolith mode in a split database scenario -- in this case, the silo mode does not help
+    select the correct implementation to ensure non mingled transactions.  This helper picks a backing implementation
+    by checking if an open transaction exists for the routing of the given model for a backend implementation.
+
+    If no transactions are open, it uses a given default implementation instead.
+    """
+
+    _constructors: Mapping[Type[Model], Callable[[], ServiceInterface]]
+    _default: Callable[[], ServiceInterface]
+
+    def __init__(
+        self,
+        mapping: Mapping[Type[Model], Callable[[], ServiceInterface]],
+        default: Callable[[], ServiceInterface],
+    ):
+        self._constructors = mapping
+        self._default = default
+
+    def __getattr__(self, item: str) -> Any:
+        for model, constructor in self._constructors.items():
+            if in_test_environment():
+                from sentry.testutils.hybrid_cloud import simulated_transaction_watermarks
+
+                open_transaction = (
+                    simulated_transaction_watermarks.connection_transaction_depth_above_watermark(
+                        using=router.db_for_write(model)
+                    )
+                    > 0
+                )
+            else:
+                open_transaction = transaction.get_connection(
```

Comment:
Can this branch be tested?

Reply:
I can add a test for this branch I think, yes.

Reply:
Added